### PR TITLE
Automated market resolution

### PR DESCRIPTION
**Pull request by TrueMilli**
### This feature will brings automatic resolutions to abandoned markets.

Automatic resolution means that unresolved markets get resolved as 'MKT' (paying out investors according to the current odds) at a specified date. By default that date is 1 week after the market closes but it can be modified by the market creator.

## Features
* Market info panel: The auto resolution time (editable) and the auto resolution 'MKT' is displayed
* Editing close date: The auto resolution time is set to close date + 1 week if it would be less otherwise 
* Contract: The auto resolution time is added
* Firestore rules: The market creator is allowed to change the auto resolution time 
* Market creation: The auto resolution time is set as close date + 1 week
* LiteMarket: Added auto resolution time and auto resolution
* Firebase: Every hour is checked whether markets should be auto resolved
* Firebase: Existing contracts get an auto resolution time (current date + 2 weeks for closed contracts, close date + 1 week otherwise)

## Already done on prod
* Update firebase rules (add autoResolutionTime to update)
* Add a close time manually to old contracts lacking it
* Run add-auto-resolution.ts (/.js) on db

## ToDo prod db
* Deploy updated version of createmarket to db
* Deploy updated version of resolveMarket to db (refactoring)
* Deploy autoResolveMarkets to db
* Reset auto resolution times (initial setup was on 2022-06-10)

## Possible features for later PRs
* eMails 1 day before auto resolve
* Notification that markets auto close now for creators with closed unresolved markets
* Change the "Market Info" interface to look nicer and allow edits of the close date
* Changeable auto resolution type (if requested)
* Removable auto resolution time
* Make close time required

## Untested
* Numeric markets are not tested because they are disabled.

### CODE REVIEW

**Comment by TrueMilli** [comment]
Re static analysis: 
I don't understand the prettier complaint, I don't see components/checkbox.tsx and hooks/use-event.ts in the changed files or even the repository.
Can somebody help me out?

File: functions/src/scripts/add-auto-resolution.ts
Comment (sipec) [inline functions/src/scripts/add-auto-resolution.ts:1]: I take mqp's position that there are particular markets that were intended by their creators to be resolved years into the future, which we'll need to manually fix. I think this is great for the majority of markets though!

**Review by sipec — COMMENTED** [review]
This looks pretty good!

File: common/contract.ts
Comment (mqp) [inline common/contract.ts:37]: This seems like a lie because there will be many contracts for which these fields are not set.

File: functions/src/resolve-market.ts
Comment (mqp) [inline functions/src/resolve-market.ts:59]: Why not query for this? Filtering them in memory seems super gross.

File: functions/src/resolve-market.ts
Comment (mqp) [inline functions/src/resolve-market.ts:51]: I would be more inclined to run this, say, every minute.

File: functions/src/resolve-market.ts
Comment (mqp) [inline functions/src/resolve-market.ts:120]: This whole file kind of seems like a mess to me, but I am not going to say you have to fix it to merge this. It would just be nice if it were nicer.

File: functions/src/scripts/add-auto-resolution.ts
Comment (mqp) [inline functions/src/scripts/add-auto-resolution.ts:23]: I think you meant to do something like a `map` and `Promise.all` here.

File: functions/src/scripts/add-auto-resolution.ts
Comment (mqp) [inline functions/src/scripts/add-auto-resolution.ts:27]: Clearly you meant `contracts`.

File: web/components/contract/contract-info-dialog.tsx
Comment (mqp) [inline web/components/contract/contract-info-dialog.tsx:188]: I don't really think that the UI where you click the pencil and it becomes a writable input has any advantages at all over the UI where it's just an input in the first place, and it has disadvantages (like complexity, probable inaccessibility, and you have to click the pencil.)

**Review by mqp — APPROVED** [review]
Cool, makes sense to me. I didn't squint at the resolution logic very hard but I assume it basically just moved around rather than changed, so it should be OK.

File: functions/src/scripts/add-auto-resolution.ts
Comment (TrueMilli) [reply functions/src/scripts/add-auto-resolution.ts:1]: I'll implement an eMail notification 1 day before the auto resolution to warn creators to change the auto resolution date in case they don't want their markets resolved yet in another PR.

We can think of other notifications for the creators as well since the feature is new.

File: functions/src/resolve-market.ts
Comment (TrueMilli) [reply functions/src/resolve-market.ts]: Fixed

File: functions/src/resolve-market.ts
Comment (TrueMilli) [reply functions/src/resolve-market.ts:51]: I thought about it but it should not be important when exactly the auto resolution happens - no one can trade anymore and +1h doesn't really make a difference after a week.

Is there anything I'm missing?

File: functions/src/scripts/add-auto-resolution.ts
Comment (TrueMilli) [inline functions/src/scripts/add-auto-resolution.ts:21]: Query for non resolved contracts above instead of this.

File: functions/src/resolve-market.ts
Comment (jahooma) [inline functions/src/resolve-market.ts:118]: There's no need to pass the contractDoc, this function can simply generate that reference.

Comment (jahooma) [inline functions/src/resolve-market.ts:118]: I'd prefer not to have this `autoResolution` param. If we can avoid having it, it makes the code simpler.

Instead, the `autoResolve` function should do any special tasks, e.g. updating the description to say it was auto resolved.

File: common/contract.ts
Comment (jahooma) [inline common/contract.ts:36]: I don't see why we want this field if auto resolve  always uses one resolution.

If there's another usecase for it, we should add it at that time. I'm also skeptical you would want any specific resolution telegraphed in advanced, other than `N/A`. If we had some api that was auto-resolving it, that should be variable, not fixed in advanced.

**Review by jahooma — CHANGES_REQUESTED** [review]
This is very cool, and a feature we may need.

I have a few concerns before this is merged.

- I'm not sure that the auto-resolution is adequately communicated to the market creator. They're not going to check the market info dialog. We probably want a label in the market creation page near the close date input that informs them the market will auto-resolve 1 week later by default. Secondly, we need an email and in-app notification that warns market creators that their contract is about to auto-resolve, probably generated 24 hours before.
- I'm very confident that auto-resolving to `MKT` is a bad idea. Leads to strong incentives for whales to camp out on these markets and bet them all the way to the opposite resolution, taking money from everyone who bet wisely. I would not be comfortable merging this unless we change it to `N/A` as the default, or you can convince me otherwise.
- I'm skeptical of other kinds of "auto" resolutions (wouldn't they just be normal resolutions but performed via api?). It doesn't make sense in general that the auto resolution is known in advance, except for `N/A`
- I'm not sure having both a close date and auto-resolution date is the right approach. I now think that we don't want markets to ever close. In that case, we might want one `resolvesInvalidTime` which is the time at which the market resolves `N/A` if the creator hasn't resolved by. In any case, it would be better to have more confidence in this approach before merging.

**Comment by jahooma** [comment]
Hi @TrueMilli,

Thanks for creating this PR! Good news —we've discussed it a bit more in our standup, and we're on board with the overall change.

- Auto-resolving to MKT is ok for now
- When the change goes in, ping me and I'll update the email that gets sent when one of your markets closes, so that it warns you that the market is set to auto-resolve.
- FYI: The current plan is to eventually deprecate `closeTime` and just have the auto-resolution time, so markets never close.
- Specific comments raised in code review still need to be addressed fully before merge. We also want the change to be thoroughly tested since you are new to the codebase.

Cheers,
James

File: web/components/contract/contract-info-dialog.tsx
Comment (TrueMilli) [reply web/components/contract/contract-info-dialog.tsx:188]: I agree in general, but I don't like the way the datepicker doesn't fit in with the other fields:
image

It's probably good to think about it some more outside of this feature. E.g. there is no reason for me the close date shouldn't be editable there as well.

File: common/contract.ts
Comment (TrueMilli) [reply common/contract.ts:36]: You are right. I was worried I would need to re-add it if I removed it earlier, but it's worse if we have bloat in our codebase and db. I'll remove it tomorrow.

File: functions/src/scripts/add-auto-resolution.ts
Comment (jahooma) [inline functions/src/scripts/add-auto-resolution.ts:26]: Need an await here. Or you could return. Best is probably `contracts.map((c) => () => addAutoResolutionToContract(c))` which returns for you.

File: web/lib/firebase/contracts.ts
Comment (jahooma) [inline web/lib/firebase/contracts.ts:50]: Ideally, this would display the year if the year is not the same as our current year.

I'd either add that logic, or remove, since nothing is using the autoResolutionDate at the moment.

**Comment by TrueMilli** [comment]
> This looks pretty good — thanks for the changes you've made so far!
> 
> I left a few more comments. I think it's about ready for the next phase, where you run the script and/or and manually add a close date to contracts missing those.
> 
> Seems like it would make sense to manually add all the close dates first? Then you'd only need to run the script once.
> 
> What's the process for adding those close dates? I'd suggest
> 
> * a script to download those questions into a csv
> * use google sheets to manually set close dates for each in a column (have one of us double check the close dates)
> * download the updated csv
> * run another script to set those close dates

* All open markets on the dev db have close dates now and all unresolved markets have auto resolve dates now
* I've asked in #dev-and-api for contracts.json generated by get-json-dump.js to define close dates for the contracts lacking them

As soon as I get the file I'll define close dates and create a google doc with the suggestions. When they are double checked I create a script to set them.

**Comment by TrueMilli** [comment]
> > This looks pretty good — thanks for the changes you've made so far!
> > I left a few more comments. I think it's about ready for the next phase, where you run the script and/or and manually add a close date to contracts missing those.
> > Seems like it would make sense to manually add all the close dates first? Then you'd only need to run the script once.
> > What's the process for adding those close dates? I'd suggest
> > 
> > * a script to download those questions into a csv
> > * use google sheets to manually set close dates for each in a column (have one of us double check the close dates)
> > * download the updated csv
> > * run another script to set those close dates
> 
> * All open markets on the dev db have close dates now and all unresolved markets have auto resolve dates now
> * I've asked in #dev-and-api for contracts.json generated by get-json-dump.js to define close dates for the contracts lacking them
> 
> As soon as I get the file I'll define close dates and create a google doc with the suggestions. When they are double checked I create a script to set them.

Scripts created and pushed to the PR. Direct links are in the discord as well.

**Comment by jahooma** [comment]
Cool! I'll run the script later today

**Comment by jahooma** [comment]
Ok, I've run the scripts!
- All markets now have a closeTime
- (I also modified the script to set a closeTime for markets that were already resolved to the resolutionTime)
- Every unresolved market now has an autoResolutionTime

A couple thoughts
- We should make the `closeTime` field non-optional
- But, we should make the `autoResolutionTime` field optional, because I did not fill it in for markets that were already resolved. It also makes sense that we might want to support a use case of not having an auto resolution time. This will require some changes to the PR.

One the whole, this is a nice cleanup for our contracts!

File: web/components/contract/contract-info-dialog.tsx
Comment (jahooma) [reply web/components/contract/contract-info-dialog.tsx:175]: Agree their's room for improvement here, in terms of how you edit the dates and also allowing the market close to be edited here, which is fairly important.
I'm ok for that to be a followup PR.

File: functions/src/resolve-market.ts
Comment (jahooma) [inline functions/src/resolve-market.ts:144]: This is a little suspicious. Why isn't BINARY a case in the switch statement?

I think you've lost this check from before:
```
if (outcomeType === 'BINARY') {
        if (!RESOLUTIONS.includes(outcome))
          return { status: 'error', message: 'Invalid outcome' }
}
```
I would be very careful about refactoring code like this, especially if you were not the one that wrote it initially.

**Review by jahooma — COMMENTED** [review]
A few comments from another pass on the code!

File: functions/src/resolve-market.ts
Comment (TrueMilli) [reply functions/src/resolve-market.ts:144]: I assumed `!RESOLUTIONS.includes(outcome)` was always false.

Outcome is of type resolution:
 ```
data: {
    outcome: resolution
    value?: number
    contractId: string
    probabilityInt?: number
    resolutions?: { [outcome: string]: number }
}
```

with `resolution` and `RESOLUTIONS` being defined as.

```
export type resolution = 'YES' | 'NO' | 'MKT' | 'CANCEL'
export const RESOLUTIONS = ['YES', 'NO', 'MKT', 'CANCEL'] as const
```

File: common/contract.ts
Comment (TrueMilli) [reply common/contract.ts:36]: The field being optional is definitely the right decision even if we are not able to remove the auto resolution date on the website yet. I didn't know how to make the date optional without adding another button.

Removing the date is nicer than having to set it to the year 9999.

File: functions/src/resolve-market.ts
Comment (jahooma) [reply functions/src/resolve-market.ts:145]: That's true — you're normally right, but this needs to be robust to rouge client calls. They might not abide by the required type, so we have to check all the parameters we have been passed to make sure they are correct.

File: common/contract.ts
Comment (jahooma) [reply common/contract.ts:36]: I see, yes. So you need to update the type here to add the '?' and make it optional.

**Review by jahooma — CHANGES_REQUESTED** [review]
Great. I'll approve and merge this once the comments are all addressed. Thanks!

File: functions/src/resolve-market.ts
Comment (jahooma) [inline functions/src/resolve-market.ts:158]: This is incorrect if `probabilityInt` is 0. Switch back to checking `probabilityInt !== undefined`

File: web/components/contract/contract-info-dialog.tsx
Comment (jahooma) [inline web/components/contract/contract-info-dialog.tsx:220]: I think you need to call `Date.now()`. Although, closeTime is always defined at the moment.

**Review by jahooma — APPROVED** [review]
Found a few more things to fix, but otherwise approved!

I'll rerun the script to update `autoResolutionTime` for all contracts, and merge this PR in — hopefully later today if you make the small fixes suggested.

Thanks!

File: functions/src/create-contract.ts
Comment (jahooma) [inline functions/src/create-contract.ts:119]: This is not in the right order, `autoResolutionTime` should be the last parameter. (To make this more robust, we could switch to taking all the parameters as an object instead of a param list. Not sure if that should be part of this change.)

File: functions/src/resolve-market.ts
Comment (jahooma) [inline functions/src/resolve-market.ts:63]: closeTime should be '<' Date.now() if the market is closed!

File: functions/src/resolve-market.ts
Comment (jahooma) [inline functions/src/resolve-market.ts:64]: Firebase does not allow two queries with not equal operators. Advice is to move the closeTime query into a filter operation after loading the contracts.

File: functions/src/resolve-market.ts
Comment (jahooma) [inline functions/src/resolve-market.ts:57]: Needs `.runWith({ secrets: ['MAILGUN_KEY'] })` since it sends emails.

**Review by jahooma — COMMENTED** [review]
The PR was reverted because of a few bugs I found.

**Pull request merged**
